<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Посмотрим-на-каждую-таблицу-и-посчитаем-кол-во-строк-в-них" data-toc-modified-id="Посмотрим-на-каждую-таблицу-и-посчитаем-кол-во-строк-в-них-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Посмотрим на каждую таблицу и посчитаем кол-во строк в них</a></span><ul class="toc-item"><li><span><a href="#таблица-ratings" data-toc-modified-id="таблица-ratings-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>таблица ratings</a></span></li><li><span><a href="#таблица-books" data-toc-modified-id="таблица-books-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>таблица books</a></span></li><li><span><a href="#таблица-authors" data-toc-modified-id="таблица-authors-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>таблица authors</a></span></li><li><span><a href="#таблица-reviews" data-toc-modified-id="таблица-reviews-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>таблица reviews</a></span></li><li><span><a href="#таблица-publishers" data-toc-modified-id="таблица-publishers-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>таблица publishers</a></span></li></ul></li><li><span><a href="#Посчитаем,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитаем,-сколько-книг-вышло-после-1-января-2000-года;-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Посчитаем, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку;-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так--исключим-из-анализа-брошюры;" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так--исключим-из-анализа-брошюры;-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так  исключим из анализа брошюры;</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитываем-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитываем-только-книги-с-50-и-более-оценками;-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></div>

# Описание проекта:

Компания приобрела сервис по чтению книг по подписке. Основная задача - проанализировать имеющуюся БД.

В распоряжении 5 таблиц:

 - books
 - authors
 - publishers
 - ratings
 - reviews
 
Необходимо изучить данные и сформить первичное представление о том, какими книгами и данными располагает наш сервис.

### Посмотрим на каждую таблицу и посчитаем кол-во строк в них

#### таблица ratings

In [2]:
ratings = 'SELECT * FROM ratings LIMIT 5'
get_sql_data(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [3]:
rows = ['ratings']
for table in rows:
    query = f"SELECT COUNT(*) as row_count FROM {table}"
get_sql_data(query)

row_count
0       6456

Содержит 6456 строк и данные о рейтингах:

 - rating_id — идентификатор оценки;
 - book_id — идентификатор книги;
 - username — имя пользователя, оставившего оценку;
 - rating — оценка книги.

#### таблица books

In [4]:
books = 'SELECT * FROM books LIMIT 5'
get_sql_data(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
rows = ['books']

for table in rows:
    query = f"SELECT COUNT(*) as row_count FROM {table}"
get_sql_data(query)

row_count
0       1000

Содержит 1000 строк и данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

#### таблица authors

In [6]:
authors = 'SELECT * FROM authors LIMIT 5'
get_sql_data(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
rows = ['authors']

for table in rows:
    query = f"SELECT COUNT(*) as row_count FROM {table}"
get_sql_data(query)

row_count
0        636

Содержит 636 строк и данные об авторах:

 - author_id — идентификатор автора;
 - author — имя автора.

#### таблица reviews

In [8]:
reviews = 'SELECT * FROM reviews LIMIT 5'
get_sql_data(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [9]:
rows = ['reviews']

for table in rows:
    query = f"SELECT COUNT(*) as row_count FROM {table}"
get_sql_data(query)

row_count
0       2793

Содержит 2793 строк и данные о рецензиях:

 - review_id — идентификатор обзора;
 - book_id — идентификатор книги;
 - username — имя пользователя, написавшего обзор;
 - text — текст обзора.

#### таблица publishers

In [10]:
publishers = 'SELECT * FROM publishers LIMIT 5'
get_sql_data(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
rows = ['publishers']

for table in rows:
    query = f"SELECT COUNT(*) as row_count FROM {table}"
get_sql_data(query)

row_count
0        340

Содержит 340 и данные об издательствах:
    
 - publisher_id — идентификатор издательства;
 - publisher — название издательства;

### Посчитаем, сколько книг вышло после 1 января 2000 года;

In [12]:
book_published = "SELECT COUNT(*) as book_count \
FROM books \
WHERE publication_date > '2000-01-01'"
get_sql_data(book_published)

book_count
0         819

комментарий:
 - С 1 января 2020 года было выпущено 819 книг

### Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [13]:
book_rate = 'SELECT b.title,b.book_id, COUNT(DISTINCT e.review_id) AS review_count, AVG(r.rating) AS average_rating \
FROM books AS b \
LEFT JOIN reviews AS e ON b.book_id = e.book_id \
LEFT JOIN ratings AS r ON b.book_id = r.book_id \
GROUP BY b.book_id,b.title \
ORDER BY review_count DESC'

get_sql_data(book_rate)

title  book_id  review_count  \
0                              Twilight (Twilight  #1)      948             7   
1                                  Water for Elephants      963             6   
2                                     The Glass Castle      734             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302             6   
4    The Curious Incident of the Dog in the Night-Time      695             6   
..                                                 ...      ...           ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83             0   
996                            The Natural Way to Draw      808             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672             0   
998                          Essential Tales and Poems      221             0   
999  Disney's Beauty and the Beast (A Little Golden...      191             0   

     average_rating  
0          3.662500  
1          3.977273  
2          4.206897  
3          4.414634  
4          4.081081  
..              ...  
995        3.666667  
996        3.000000  
997        5.000000  
998        4.000000  
999        4.000000  

[1000 rows x 4 columns]

комментарий:
 - Самое большое кол-во обзоров у книги Twilight, но средний рейтинг не самый выскоий.

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так  исключим из анализа брошюры;

In [14]:
pub = 'SELECT p.publisher,COUNT(title) AS books_qantity \
FROM publishers AS p \
JOIN books AS b ON b.publisher_id = p.publisher_id \
WHERE b.num_pages > 50 \
GROUP BY p.publisher \
ORDER BY books_qantity DESC \
LIMIT 1;'
get_sql_data(pub)


publisher  books_qantity
0  Penguin Books             42

комментарий:
 - Penguin Books выпустило больше всего книг толще 50 страниц.Всего 42.

### Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками;

In [15]:
great_author_1 = 'SELECT q.author AS author, AVG(q.average_rating) AS average_rating \
FROM ( \
SELECT a.author AS author,b.book_id AS book_id,AVG(r.rating) AS average_rating,COUNT(r.rating) AS cnt_rating \
FROM books AS b \
JOIN ratings AS r ON r.book_id = b.book_id \
JOIN authors AS a ON b.author_id = a.author_id \
GROUP BY a.author,b.book_id \
HAVING COUNT(r.rating_id) >= 50 \
) AS q \
GROUP BY q.author \
ORDER BY average_rating DESC \
LIMIT 1'

get_sql_data(great_author_1)

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.283844

комментарий:
 - J.K. Rowling/Mary GrandPré	 имеет самую высокий средний рейтинг в 4.28 с учетом 50 и более оценок на ее книги

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [16]:
great_users = 'SELECT AVG(review_count) AS average_review_count \
FROM ( \
    SELECT r.username, COUNT(DISTINCT r.text) AS review_count \
    FROM reviews AS r \
    JOIN ( \
        SELECT username, COUNT(*) AS rating_count \
        FROM ratings \
        GROUP BY username \
        HAVING COUNT(*) > 48 \
    ) AS ra ON r.username = ra.username \
    GROUP BY r.username \
) as subquery'

get_sql_data(great_users)

average_review_count
0                  24.0

комменатрий:
 - среднее кол-во обзоров от пользователей,которые поставили больше 48 оценок, 24 обзора

### Вывод:

- Сервис обладает набором из 1000 книг. Большинство были выпущены относительно недавно.819 выпущенных с 1 января 2000 года предлагается в сервисе, значит сервис ориентируется на современную литературу. 
- Каждая из 1000 книг имеет обзор со стороны пользователей и оценки, что может помочь потенциальным покупателям в выборе книги
- Высоко оценный писатель в сервисе, это J.K. Rowling/Mary GrandPré	у нее средняя оценка 4.28
- Были выбраны пользователи, которые поставили более 48 оценок книгам, среди них было рассчитано среднее число обзоров, которые они делают, получилось 24,3 обзора на такого пользователя, их можно отнести к наиболее активным читателям в разрезе данного датасет

На основании проведенного анализа можно сказать, что в сервисе достаточно активные пользователи, несмотря на относительно небольшую подборку книг. С учетом просмотренных рейтингов можно составить портрет пользователей, и пополнять сервис книгами от соответствующих авторов или новых, но с похожим жанром или стилем написания